File: scripts/codex_neutral_name.py
#!/usr/bin/env python3
"""SD-32 `decisions.md §24` — the Codex-generated neutral name for units
whose *own name* is Product Identity (`ability`/`deity`/`class_feature`;
population re-derived per `§17a`, do not trust a stale count without
re-running the ingest that consumes this module).

**Binding condition 1 (`§24b`-1): derived ONLY from non-PI coordinates.**
`(kind, book, source_file, source_line)` — the exact `(book, source_basename,
source_line)` join key `scripts/shape_ledger.py` already uses, plus `kind`
(`§24a`: "and optionally the record's kind"). **Never** from the PI name —
not transformed, not truncated, not hashed.

This is enforced structurally, not just by discipline: every public function
below takes `kind`/`book`/`source_file`/`source_line` and NO `name`, `key`,
free-text, or hash-of-name parameter. There is no argument through which a
PI string could enter the derivation, so "the generator's output is
unchanged when the PI name is replaced with a different string" holds by
construction. `scripts/tests/test_codex_neutral_name.py` proves it anyway,
black-box, by calling this module with two records that share coordinates
and differ only in a (never-passed) original name.

**Binding condition 6: determinism.** Pure string composition of the four
inputs — no randomness, no wall-clock, no hashing. `§24a` itself argues
against a hash ("a hash in particular is a fingerprint that confirms a
guess") in favour of a plainly-derived, self-describing name, so this
module deliberately does NOT hash anything, including the coordinates
themselves.
"""
from __future__ import annotations

import os
import re

# The literal marker every renamed record's `codex_generated_name` field
# carries `True` under, and the human-readable name prefix — both chosen so
# a reader (or a future generator) can recognise a Codex-minted identity on
# sight, per `§24b`-3 ("a field marks it... so no reader or player mistakes
# it for the printed name").
NAME_PREFIX = "Codex-Named Unit"


def _slug(value: str) -> str:
    """Lowercase, `[a-z0-9]+`-runs-joined-by-`_` normalisation. Applied only
    to non-PI coordinate strings (book names, filenames, kind names) — never
    to a record's display name or description."""
    slug = re.sub(r"[^a-z0-9]+", "_", value.lower()).strip("_")
    return slug or "x"


def neutral_coordinate_id(kind: str, book: str, source_file: str, source_line: int) -> str:
    """The stable, PI-free identifier for one unit: `kind_book_basename_line`.
    Two different units can never collide here unless they share the exact
    same `(book, source_file, source_line)` citation, which `v06_work_inventory`'s
    own enumeration already treats as the same object."""
    basename = os.path.basename(source_file)
    return f"{_slug(kind)}_{_slug(book)}_{_slug(basename)}_{int(source_line)}"


def neutral_name(kind: str, book: str, source_file: str, source_line: int) -> str:
    """The Codex-generated display name a renamed record's `data.name`
    carries. Self-describing as generated (contains `NAME_PREFIX`) so even a
    reader who never checks the `codex_generated_name` field sees it is not
    a printed name."""
    coord_id = neutral_coordinate_id(kind, book, source_file, source_line)
    return f"{NAME_PREFIX} ({coord_id})"


def neutral_key(kind: str, book: str, source_file: str, source_line: int) -> str:
    """Same identity, used wherever a record needs a `key` distinct in
    principle from `name` (mirrors the pre-existing generators' key/name
    split) — here the two are the same string, since both would otherwise
    just re-derive the identical coordinate id."""
    return neutral_name(kind, book, source_file, source_line)


def divergence_entry(kind: str, book: str, source_file: str, source_line: int, reason: str) -> dict:
    """`§22`'s divergence-recording, refined by `§24b`-4: record **that** a
    rename happened, its coordinates, and why — **never the original
    string**. This is the one shape every renamed-unit divergence log in
    this bundle must use, so a grep for a PI term across those logs can
    never find one (there is no field here it could be written into)."""
    return {
        "kind": kind,
        "book": book,
        "source_file": os.path.basename(source_file),
        "source_line": int(source_line),
        "codex_name": neutral_name(kind, book, source_file, source_line),
        "reason": reason,
    }

File: src/rules_core/codex_neutral_name.rs
//! SD-32 `decisions.md §24` — the Codex-generated neutral name for units
//! whose *own name* is Product Identity (`ability`/`deity`/`class_feature`).
//! This is the Rust port of `scripts/codex_neutral_name.py`, used by
//! generators that already run in Rust (`cache_gen::class_feature`) rather
//! than as a Python post-pass, kept byte-identical in derivation to the
//! Python module so a record renamed by either path is indistinguishable.
//!
//! **Binding condition 1 (`§24b`-1): derived ONLY from non-PI coordinates.**
//! `(kind, book, source_file, source_line)` — the exact join key
//! `scripts/shape_ledger.py` already uses, plus `kind`. **Never** from the
//! PI name — not transformed, not truncated, not hashed. Enforced
//! structurally: every public function below takes `kind`/`book`/
//! `source_file`/`source_line` and no `name`/`key`/free-text parameter at
//! all, so there is no argument through which a PI string could enter the
//! derivation. `src/rules_core/tests` (this module's own `#[cfg(test)]`
//! block) proves the black-box claim: two calls that share coordinates
//! produce the same output regardless of what original name the CALLER
//! happens to hold (never passed in).
//!
//! **Binding condition 6: determinism.** Pure string composition of the
//! four inputs — no randomness, no wall-clock, no hashing. `§24a` argues
//! against a hash ("a hash in particular is a fingerprint that confirms a
//! guess"), so this module deliberately does not hash anything, including
//! the coordinates themselves.

use std::path::Path;

/// The literal marker every renamed record's human-readable name carries —
/// chosen so a reader (or a future generator) recognises a Codex-minted
/// identity on sight, per `§24b`-3.
pub const NAME_PREFIX: &str = "Codex-Named Unit";

/// Lowercase, `[a-z0-9]+`-runs-joined-by-`_` normalisation. Applied only to
/// non-PI coordinate strings (book names, filenames, kind names) — never to
/// a record's display name or description.
fn slug(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    let mut last_was_sep = true; // suppress a leading separator
    for ch in value.chars() {
        let lower = ch.to_ascii_lowercase();
        if lower.is_ascii_alphanumeric() {
            out.push(lower);
            last_was_sep = false;
        } else if !last_was_sep {
            out.push('_');
            last_was_sep = true;
        }
    }
    while out.ends_with('_') {
        out.pop();
    }
    if out.is_empty() {
        "x".to_string()
    } else {
        out
    }
}

/// The stable, PI-free identifier for one unit: `kind_book_basename_line`.
/// Two different units can never collide here unless they share the exact
/// same `(book, source_file, source_line)` citation, which the census's own
/// enumeration already treats as the same object.
pub fn neutral_coordinate_id(kind: &str, book: &str, source_file: &str, source_line: u32) -> String {
    let basename = Path::new(source_file)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or(source_file);
    format!("{}_{}_{}_{}", slug(kind), slug(book), slug(basename), source_line)
}

/// The Codex-generated display name a renamed record's `data.name` carries.
/// Self-describing as generated ([`NAME_PREFIX`] prefix) so even a reader who
/// never checks the `codex_generated_name` field sees it is not a printed
/// name.
pub fn neutral_name(kind: &str, book: &str, source_file: &str, source_line: u32) -> String {
    format!("{NAME_PREFIX} ({})", neutral_coordinate_id(kind, book, source_file, source_line))
}

/// Same identity, used wherever a record needs a `key` distinct in
/// principle from `name` — here the two are the same string, mirroring
/// `scripts/codex_neutral_name.py::neutral_key`.
pub fn neutral_key(kind: &str, book: &str, source_file: &str, source_line: u32) -> String {
    neutral_name(kind, book, source_file, source_line)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// `§24b`-1's own required proof: the generator's output is unchanged
    /// when the (never-passed) PI name is swapped for a different string.
    /// There is no name parameter to swap, so this test proves the claim by
    /// construction: two callers with different "original name" context
    /// (never threaded into these functions at all) produce identical
    /// output for the same coordinates.
    #[test]
    fn output_is_unchanged_when_the_pi_name_is_swapped_for_a_different_string() {
        // Caller A "believes" the original name is one string; caller B
        // believes it is a completely different one. Neither belief has any
        // parameter to travel through -- both derive from the SAME call.
        let original_name_a = "Iomedae"; // stand-in PI-shaped string, not asserted to be real PI
        let original_name_b = "Some Entirely Different Name";
        assert_ne!(original_name_a, original_name_b);

        let a = neutral_name("deity", "inner_sea_gods", "isg_deities.lst", 42);
        let b = neutral_name("deity", "inner_sea_gods", "isg_deities.lst", 42);
        assert_eq!(a, b, "same coordinates must produce the same name regardless of any external name context");
        assert!(!a.contains(original_name_a) && !a.contains(original_name_b));
    }

    #[test]
    fn coordinate_id_is_pure_composition_of_kind_book_basename_line() {
        let id = neutral_coordinate_id("class_feature", "adventurers_guide", "support/ag_abilities_class_2.lst", 913);
        assert_eq!(id, "class_feature_adventurers_guide_ag_abilities_class_2_lst_913");
    }

    #[test]
    fn determinism_two_calls_same_inputs_byte_identical() {
        let n1 = neutral_name("deity", "book", "file.lst", 7);
        let n2 = neutral_name("deity", "book", "file.lst", 7);
        assert_eq!(n1, n2);
        let k1 = neutral_key("deity", "book", "file.lst", 7);
        let k2 = neutral_key("deity", "book", "file.lst", 7);
        assert_eq!(k1, k2);
    }

    #[test]
    fn different_coordinates_never_collide() {
        let a = neutral_coordinate_id("deity", "book", "file.lst", 7);
        let b = neutral_coordinate_id("deity", "book", "file.lst", 8);
        assert_ne!(a, b);
    }

    #[test]
    fn slug_never_produces_an_empty_or_leading_underscore_segment() {
        assert_eq!(slug(""), "x");
        assert_eq!(slug("!!!"), "x");
        assert_eq!(slug("  Hello, World!  "), "hello_world");
    }
}
